lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,426 @@
+  ///         TransitionPhase.HIDDEN: domProps()..className = 'hidden',
+  ///       })
+  ///     )(
+  ///       // The child that has CSS transitions
+  ///     );
+  ///   },
+  ///   $WithTransitionExampleConfig, // ignore: undefined_identifier
+  /// );
+  /// ```
+  Map<TransitionPhase, Map> childPropsByPhase;
+
+  /// The amount of time to wait for a CSS transition to complete before assuming something went wrong
+  /// and canceling the subscription.
+  ///
+  /// If the CSS transition duration on the node is longer than `1s`, you will need to set this to a Duration
+  /// that is greater than or equal to the expected CSS duration of the node.
+  ///
+  /// > Default `const Duration(seconds: 1)`
+  Duration transitionTimeout;
+}
+
+class WithTransitionProps = UiProps with v2.TransitionPropsMixin, WithTransitionPropsMixin;
+
+mixin WithTransitionState on UiState {
+  /// The current transition phase, derived from the current [WithTransitionPropsMixin.isShown] value.
+  ///
+  /// Do not set directly.
+  @protected
+  TransitionPhase $transitionPhase;
+}
+
+class WithTransitionComponent extends UiStatefulComponent2<WithTransitionProps, WithTransitionState> {
+  /// Whether the overlay is not guaranteed to transition in response to the current
+  /// state change.
+  ///
+  /// _Stored as variable as workaround for not adding breaking change to [_stateToBeginHiding] API._
+  ///
+  /// A transition may not always occur when the state moves from SHOWING to HIDING
+  /// if the PRE_SHOWING-->SHOWING-->HIDING transition happens back-to-back.
+  ///
+  /// Better to not always transition when the user is ninja-toggling a transitionable
+  /// component than to break state changes waiting for a transition that will never happen.
+  var _transitionNotGuaranteed = false;
+
+  /// Timer used to determine if a transition timeout has occurred.
+  Timer _transitionEndTimer;
+
+  /// Stream for listening to `transitionend` events on the node.
+  StreamSubscription _endTransitionSubscription;
+
+  final _transitionNodeRef = createRef<Element>();
+
+  @override
+  get defaultProps => (newProps()
+    ..addProps(v2.TransitionPropsMixin.defaultProps)
+    ..isShown = false
+    ..transitionTimeout = const Duration(seconds: 1)
+    ..childPropsByPhase = const {
+      v2.TransitionPhase.PRE_SHOWING: {},
+      v2.TransitionPhase.SHOWING: {},
+      v2.TransitionPhase.SHOWN: {},
+      v2.TransitionPhase.HIDING: {},
+      v2.TransitionPhase.HIDDEN: {},
+    }
+  );
+
+  @override
+  get initialState => (newState()
+    ..$transitionPhase = _transitionPhaseDerivedFromProps(props)
+  );
+
+  @override
+  getDerivedStateFromProps(Map nextProps, Map prevState) {
+    final tNextProps = typedPropsFactory(nextProps);
+    final tPrevState = typedStateFactory(prevState);
+    if (tNextProps.isShown && _isOrWillBeHidden(tPrevState.$transitionPhase)) {
+      return _stateToBeginShowing(tNextProps, tPrevState.$transitionPhase);
```

Comment:
getDerivedSTateFromProps shouldn't have side effects (e.g., calling a props callback in this case) since in concurrent mode this method can be called multiple times without actually rendering the result. Can we move those to `componentDidUpdate`?